Compilation failure on aarch64 with Arm Compiler for Linux's flang

Using Spack to set up config and compile, the following error occurs: 5 errors found in build log.

It makes no difference if I disable the non-standard float16 - using an arg to cmake of: `-DHDF5_ENABLE_NONSTANDARD_FEATURE_FLOAT16:BOOL=OFF`


```
5 errors found in build log:
     3047    F90-S-0081-Illegal selector - KIND parameter has unknown value for data type  (/tmp/user/spack-stage/s
             pack-stage-hdf5-1.14.5-rzhstpbozkrcypkkfn5ntevpg2gwfof7/spack-build-rzhstpb/fortran/static/H5_gen.F90: 604
             9)
     3048      0 inform,   0 warnings,   1 severes, 0 fatal for h5pget_kind_16
     3049    F90-S-0081-Illegal selector - KIND parameter has unknown value for data type  (/tmp/user/spack-stage/s
             pack-stage-hdf5-1.14.5-rzhstpbozkrcypkkfn5ntevpg2gwfof7/spack-build-rzhstpb/fortran/static/H5_gen.F90: 609
             7)
     3050      0 inform,   0 warnings,   1 severes, 0 fatal for h5pregister_kind_16
     3051    F90-S-0081-Illegal selector - KIND parameter has unknown value for data type  (/tmp/user/spack-stage/s
             pack-stage-hdf5-1.14.5-rzhstpbozkrcypkkfn5ntevpg2gwfof7/spack-build-rzhstpb/fortran/static/H5_gen.F90: 614
             5)
     3052      0 inform,   0 warnings,   1 severes, 0 fatal for h5pinsert_kind_16
  >> 3053    make[2]: *** [fortran/src/CMakeFiles/hdf5_fortran-static.dir/build.make:325: fortran/src/CMakeFiles/hdf5_f
             ortran-static.dir/__/static/H5_gen.F90.o] Error 1
     3054    make[2]: Leaving directory '/tmp/user/spack-stage/spack-stage-hdf5-1.14.5-rzhstpbozkrcypkkfn5ntevpg2gw
             fof7/spack-build-rzhstpb'
```
This is using Arm Compiler for Linux - which is LLVM derivative but the NVIDIA/PGI fortran Flang.

As the float16 is disabled, it ought not to be trying to generate it?

--- a/m4/aclocal_fc.f90
+++ b/m4/aclocal_fc.f90
@@ -183,8 +183,11 @@ END PROGRAM FC_AVAIL_KINDS
 PROGRAM FC08_AVAIL_KINDS
       USE, INTRINSIC :: ISO_FORTRAN_ENV, ONLY : stdout=>OUTPUT_UNIT, integer_kinds, real_kinds, logical_kinds
       IMPLICIT NONE
-      INTEGER :: ik, jk, k, max_decimal_prec
+      INTEGER :: ik, jk, k, kk, max_decimal_prec
       INTEGER :: num_rkinds, num_ikinds, num_lkinds
+      LOGICAL :: found_rkinds( 1:SIZE(real_kinds) )
+      CHARACTER(LEN=1) :: sep
+
 
       ! Find integer KINDs
 
@@ -203,28 +206,33 @@ PROGRAM FC08_AVAIL_KINDS
 
       num_rkinds = SIZE(real_kinds)
 
+      ! some compilers (ACfL 24) reported REAL=16 kind, but refused to
+      ! compile with it.  Verify the kind can be selected in SELECTED_REAL_KIND.
+      found_rkinds(:) = .FALSE.
       max_decimal_prec = 1
 
       prec: DO ik = 2, 36
          exp: DO jk = 1, 700
             k = SELECTED_REAL_KIND(ik,jk)
             IF(k.LT.0) EXIT exp
+            DO kk = 1,num_rkinds
+               IF (real_kinds(kk) == k) found_rkinds(kk) = .TRUE.
+            ENDDO
             max_decimal_prec = ik
          ENDDO exp
       ENDDO prec
 
+      sep = ""
       DO k = 1, num_rkinds
-         WRITE(stdout,'(I0)', ADVANCE='NO') real_kinds(k)
-         IF(k.NE.num_rkinds)THEN
-            WRITE(stdout,'(A)',ADVANCE='NO') ','
-         ELSE
-            WRITE(stdout,'()')
-         ENDIF
+         IF(.NOT. found_rkinds(k)) CYCLE
+         WRITE(stdout,'(A,I0)', ADVANCE='NO') TRIM(sep), real_kinds(k)
+         sep = ","
       ENDDO
+      WRITE(stdout,'()')
 
      WRITE(stdout,'(I0)') max_decimal_prec
      WRITE(stdout,'(I0)') num_ikinds
-     WRITE(stdout,'(I0)') num_rkinds
+     WRITE(stdout,'(I0)') COUNT(found_rkinds)
 
      ! Find logical KINDs
 

